### PR TITLE
remove Dockerfile, use only pure go dabase backends

### PR DESCRIPTION
**Pull request by faddat**
This PR:

* updates golang to v1.22.* and pebble v1.1.0
* allows cometbft downstream to reduce its usage of Docker
* reduces CI time on this repository by over 95%
* keeps goleveldb for backwards compatiblity, and badger because it's pure go
* eliminates unmaintained database backends with no future:
  * bolt
* eliminates database backends that complicate both cometbft-db and cometbft by forcing the use of cgo and docker:
  * rocksdb / grocksdb
  * cleveldb  
* eliminates the need for build tags

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -11,11 +11,27 @@ jobs:
...
     runs-on: ubuntu-latest
-    container: cometbft/cometbft-db-testing
     steps:
       - uses: actions/checkout@v4
-      - uses: golangci/golangci-lint-action@v4.0.0
-        with:
-          args: --timeout 10m
-          version: latest
-          github-token: ${{ secrets.github_token }}
+      runs-on: ubuntu-latest
```
Comment (melekes) [inline .github/workflows/lint.yml:16]: What's the reason for running it in Docker?

Comment (faddat) [reply .github/workflows/lint.yml:16]: make it similar to the tests.  It was run in docker previously, too.  We could also run it outside of docker, but we would need to change the pattern where we use build tags to specify which DB we are using.  

If you'd like to do that, and reduce DBs to only those that work I can whip up that PR with ease.

Comment (faddat) [reply .github/workflows/lint.yml:16]: ...sir, I am going to just do that PR.  It is really the sensible way to cut down this repository to the only two reasonable options, and we've done the user surveys

**Comment by faddat** [comment]
@melekes -- this is my new proposed PR.  It is simpler, easier to maintain, and faster all around.

**Comment by faddat** [comment]
@melekes -- this is my new proposed PR.  It is simpler, easier to maintain, and faster all around.

**Comment by faddat** [comment]
When this PR started, tests took more than ten minutes.  They now take less than one minute.

**Comment by melekes** [comment]
Why keep goleveldb if it's unmaintained?

**Comment by faddat** [comment]
@melekes -- backwards compatibility for in production chains. 

Unfortunately I think we will want to support it for some time to come.

**Comment by faddat** [comment]
@melekes - this is now being marked as stale, do you reckon we want to use it?

**Comment by melekes** [comment]
I agree with the general direction 👍 and would like to see this merged.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -27,17 +27,13 @@ jobs:
...
-      - name: Build Docker image
-        uses: docker/build-push-action@v5
+      - uses: actions/setup-go@v4
         with:
-          context: ./tools
-          file: ./tools/Dockerfile
-          tags: "cometbft/cometbft-db-testing:latest"
-          load: true
+          go-version: '1.22'
+          cache: false
```
Comment (melekes) [inline .github/workflows/ci.yml:33]: why cache is false?

File: .github/workflows/govulncheck.yml [hunk 3]
```diff
@@ -18,16 +18,8 @@ jobs:
     steps:
       - uses: actions/setup-go@v5
         with:
-          go-version: "1.21"
+          go-version: "1.22"
           check-latest: true
       - uses: actions/checkout@v4
-      - uses: technote-space/get-diff-action@v6
```
Comment (melekes) [inline .github/workflows/govulncheck.yml:24]: why remove?

File: .github/workflows/lint.yml [hunk 4]
```diff
@@ -1,21 +1,55 @@
-name: Lint
+name: golangci-lint
 on:
   push:
     branches:
+      - master
```
Comment (melekes) [inline .github/workflows/lint.yml:5]: ??

File: Makefile [hunk 5]
```diff
@@ -78,15 +48,7 @@ docker-test-image:
 .PHONY: docker-test-image
 
 # Runs the same test as is executed in CI, but locally.
```
Comment (melekes) [inline Makefile:50]:
```suggestion
```

File: goleveldb.go [hunk 6]
```diff
@@ -28,7 +28,18 @@ func NewGoLevelDB(name string, dir string) (*GoLevelDB, error) {
 }
 
 func NewGoLevelDBWithOpts(name string, dir string, o *opt.Options) (*GoLevelDB, error) {
-	dbPath := filepath.Join(dir, name+".db") //nolint:goconst
+	if o == nil {
+		o = &opt.Options{
+			WriteBuffer:                   64 * opt.MiB,
```
Comment (melekes) [inline goleveldb.go:33]: Where did this come from?

File: pebble.go [hunk 7]
```diff
@@ -32,6 +29,10 @@ func NewPebbleDB(name string, dir string) (*PebbleDB, error) {
 }
 
 func NewPebbleDBWithOpts(name string, dir string, opts *pebble.Options) (*PebbleDB, error) {
+	// configure pebble
+	opts.MemTableSize = 256 << 20           // 256 MB
```
Comment (melekes) [inline pebble.go:33]: this should be a separate PR. Ideally 3 PRs: 1) updating Go 2) Removing dbs other than goleveldb / pebbledb 3) configuring default options

File: pebble.go [hunk 8]
```diff
@@ -186,8 +187,27 @@ func (db *PebbleDB) Print() error {
 }
 
 // Stats implements DB.
-func (*PebbleDB) Stats() map[string]string {
-	return nil
+func (db *PebbleDB) Stats() map[string]string {
+	stats := make(map[string]string)
```
Comment (melekes) [inline pebble.go:191]: same. ideally a separate PR

File: pebble.go [hunk 9]
```diff
@@ -412,13 +428,23 @@ func (itr *pebbleDBIterator) Value() []byte {
 }
 
 // Next implements Iterator.
-func (itr pebbleDBIterator) Next() {
-	itr.assertIsValid()
+// Next implements Iterator.
```
Comment (melekes) [inline pebble.go:431]:
```suggestion
```

File: pebble.go [hunk 10]
```diff
@@ -412,13 +428,23 @@ func (itr *pebbleDBIterator) Value() []byte {
...
+
 	if itr.isReverse {
 		itr.source.Prev()
 	} else {
 		itr.source.Next()
 	}
+
+	// After moving the iterator, we need to check if it's still valid.
+	if !itr.source.Valid() {
+		itr.isInvalid = true
```
Comment (melekes) [inline pebble.go:446]: Is this fixing some bug?

File: pebble_test.go [hunk 11]
```diff
@@ -1,11 +1,10 @@
-//go:build pebbledb
```
Comment (melekes) [inline pebble_test.go:1]: why remove build tags?

**Comment by melekes** [comment]
> eliminates unmaintained database backends with no future:
bolt
badger

badger is very much maintained https://github.com/dgraph-io/badger

**Comment by faddat** [comment]
@melekes it is pure go, too, so would you like to keep it?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (faddat) [reply .github/workflows/ci.yml:33]: caching tends to cause issues and this is very fast anyhow

File: .github/workflows/govulncheck.yml (hunk 3 above)
Comment (faddat) [reply .github/workflows/govulncheck.yml:24]: only useful when things are crazy slow, and they're no longer crazy slow.

**Comment by faddat** [comment]
oh, sorry for the delay here and thanks for the review!  I think I have the branch that fixes all this on my machine.

File: .github/workflows/govulncheck.yml (hunk 3 above)
Comment (faddat) [reply .github/workflows/govulncheck.yml:24]: it really isn't needed now that ci is fast.

**Comment by faddat** [comment]
I think this is ready to rock now, but there is a valid question of weather or not to include pebble.

**Comment by faddat** [comment]
If we are going to merge this, we should.  Blocks upgrading comet to go 1.22

**Comment by adizere** [comment]
I would prefer to sunset database backends iteratively:

1. mark them as deprecated -- comet-db v0.12
    * then give it some time for users to acclimatize
3. remove them -- comet-db v0.13

I would also keep rocksdb around, since some users (at least Cronos, potentially others) clearly make use of it.

Therefore, we should separate this PR into multiple smaller PRs:
* golang to v1.22.* and pebble v1.1.0
* mark cleveldb and bold as deprecated
* Docker improvements

@faddat do you plan on helping with any of the above? if so, please open corresponding PRs. if not, we plan to work on it this week so we'll go ahead within a couple of days

**Comment by faddat** [comment]
@adizere 

it is necessary to update the dockerfile in order to update grocksdb.  So ci will take a very long time now. 

As for phases, that's fine. 

We will not be able to improve docker without removing rocksdb (and other databases that use cgo)
